### ISSUE TITLE
Revising rule(s) to protect master against accidental pushes

### ISSUE THREAD

**Issue by mikemc** [naobservatory/mgs-workflow#134]
Just now, I accidentally pushed some edits to the Readme. I thought I was working on a branch but was actually on master; when I ran `git push`, my commits were pushed, with the below git output which shows that I was able to bypass the rule that requires changes to be via a pull request.

```sh
❯ git push
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 10 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 1.05 KiB | 1.05 MiB/s, done.
Total 3 (delta 2), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
remote: Bypassed rule violations for refs/heads/master:
remote:
remote: - Cannot update this protected ref.
remote:
remote: - Changes must be made through a pull request.
remote:
To github.com:naobservatory/mgs-workflow.git
   829e2d4..a46b846  master -> master
```

I assume I was allowed to do this because I'm an admin of the NAO organization, and the the rule we have setup which requires pull requests has an exception in the "Bypass list" for the "Repository admin" role and is set to "Always allow". I guess I'm in the "Repository admin" role by virtue of being an admin of our GitHub organization. 

In this particular case, my commits were just to the readme so there's little damage (and I'm optimistic that the changes will be welcome) but I want to flag my mistake and raise the question of how we can prevent similar mistakes in the future. Some options:

1. Remove the bypass (I did this for now just to be safe) --- I still need to verify that this stops admins from accidental pushes
2. Change the bypass to something else that only applies to @willbradshaw --- instead of "Repository admin", a "Maintain" role as another option that might do the trick, though I don't know how the "Maintain" role is defined or set
3. Trusting admins to be _very careful_ --- I'll certainly be more careful now that I've made this mistake once, but this doesn't seem like the best long term option

**Comment by willbradshaw** [comment]
I agree this would be good. Looking into it now, it looks like I can only configure the bypass at the role level; i.e. there's no way to set up a bypass that wouldn't apply to all org admins.

I'll look into this more.

**Comment by willbradshaw** [comment]
OK, I've now done some further work configuring this. It should now be the case that:

- No-one, including admins, can push to master or dev without a pull request and passing tests. (i.e. there is no bypass for this rule)
- No-one except admins can push to master at all (a stricter rule preventing all pushes, with an admin bypass)

@mikemc could you do a quick test here and confirm that the thing that happened previously is now blocked?

**Comment by mikemc** [comment]
> @mikemc could you do a quick test here and confirm that the thing that happened previously is now blocked?

Confirmed!

I first tested on a fork (https://github.com/naobservatory/mgs-workflow-permissions-test). I copied the two rules over to the fork (you can manually export the rules to json files). I then tried to push a test commit and got an error as expected,

```
❯ git push 
(...)                                                                                                                                       
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
remote: error: GH013: Repository rule violations found for refs/heads/master.
remote: Review all repository rules at https://github.com/naobservatory/mgs-workflow-permissions-test/rules?ref=refs%2Fheads%2Fmaster
remote:
remote: - Changes must be made through a pull request.
remote:
remote: - 5 of 5 required status checks are expected.
remote:
To github.com:naobservatory/mgs-workflow-permissions-test.git
 ! [remote rejected] master -> master (push declined due to repository rule violations)
error: failed to push some refs to 'github.com:naobservatory/mgs-workflow-permissions-test.git'
```

I then tried pushing a test commit to the dev and master branchs on this (the main mgs-workflow) repo, and again got the expected error.

**Comment by willbradshaw** [comment]
OK great, thanks! I'm calling this complete for now.
